waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+
+          pushGroupNode childNodes
+
+          if errKindIsImmediatelyFollowing
+            then return ()
+            else throwError (Unwind kind) -- Re-throw unwind
+        ParseError perr -> do
+          -- ASSUMPTION: The token errored on has not been consumed
+          --
+          -- First, check whether the kind we errored on is in the first `pstateFollowing`
+          -- list. If it is not, we consume the token.
+          --
+          -- Then, report the error to the state. Now, we have recovered!
+          followingStack <- gets pstateFollowing
+          let errKind = case perr of
+                UnexpectedEOF _ _ -> Nothing
+                Unexpected _ k _ -> Just k
+          let errKindIsFollowingKind = any (errKind `TokenSet.member`) followingStack
+          let errKindIsImmediatelyFollowing = case followingStack of
+                [] -> False -- technically unreachable, but this is sensible (TODO: consider using error here?)
+                (following : _) -> errKind `TokenSet.member` following
+
+          -- Make the error node and possibly consume the token
+          nodeForErr <-
+            if errKindIsFollowingKind
+              then do
+                return $ errorNode perr True
+              else do
+                advance
+                return $ errorNode perr False
+
+          -- Report the error and add the error node
+          pushError perr
+          pushGroupNode $ childNodes ++ [nodeForErr]
+
+          when (errKindIsFollowingKind && not errKindIsImmediatelyFollowing) $ do
+            throwError $ Unwind errKind
+
+-- | Parse "Chain" rule. Runs the first and then the second parser.
+parseChain :: GrammarRule -> GrammarRule -> ParserM ()
+parseChain first second = do
+  consumeWhitespace
+  -- First, figure out following set for @second@
+  let following = findFirstTokens second
+  pushFollowing following
+  mbError <- tryError (doParse first)
+  -- Make sure to remove following set now that we're leaving this "Then"
+  popFollowing
+  -- Run second (or throw error if first failed)
+  case mbError of
+    Nothing -> doParse second
+    Just err -> throwError err
+
+-- | Parse "Alternative" rule. If the left parser can consume the next token in
+-- the input, run that parser. If the right parser can consume the next token,
+-- run that parser. Otherwise, report an error.
+parseAlternative :: GrammarRule -> GrammarRule -> ParserM ()
+parseAlternative left right = do
+  consumeWhitespace
+  let leftFirstTokens = findFirstTokens left
+  let rightFirstTokens = findFirstTokens right
+  nextToken <- peek
+  let nextKind = tokenKind <$> nextToken
+  if nextKind `willSucceedIn` left
+    then doParse left
+    else
+      if nextKind `willSucceedIn` right
+        then doParse right
+        else makeError nextToken (leftFirstTokens `TokenSet.union` rightFirstTokens)
+
+-- | Advance past whitespace, making nodes for the whitespace as it goes
```

Comment:
Hm but this doesn't consume just whitespace, it also consumes comments? So it cosumes "trivia", not whitespace, right?